apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,391 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Util;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// The SessionPool is the preferred method of managing Spanner Sessions.
+    /// It performs pooling of sessions with an evict timer along with configurable settings to control the maximum
+    /// number of pooled and active sessions.
+    /// To create a session using the sessionpool, use the extension method on SpannerClient.
+    ///   var session = await SpannerClient.CreateSessionFromPoolAsync(project, spannerInstance, database, cancellationtoken);
+    /// To release a session back to the sessionpool:
+    ///   session.ReleaseToPool();
+    /// If you use the SessionPool, you must make sure sessions are properly released back into the pool and are not simply GC'd.
+    /// Allowing a session to GC will incur penalties on the current process as the session will count towards the maximum allowed
+    /// and it will incur a penalty on other Spanner processes because it will be an hour before the server frees the session if 
+    /// it's not properly deleted.
+    /// </summary>
+    internal static class SessionPool
```

Comment:
Do you have a test like this:
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolStressTest.java

This essentially does lot of operations on the pool concurrently and then asserts that all the invariants are maintained. It mocks out the spanner api so it runs very fast. This would be very useful to catch subtle concurrency bugs.